[openwebnet] OpenWebNet Binding initial contribution, migration from OH1

mvalla opened the pull request:
**OpenWebNet (BTicino/Legrand) Binding**

This binding integrates BTicino/Legrand MyHOME home automation systems into openHAB (see #5195).
The binding has to be considered as an enhancement and replacement of the BTicino v1 binding.
WHOs (devices) supported in this first PR:
- WHO=1 Lighting
- WHO=13 Gateway Management

Community thread: 
https://community.openhab.org/t/bticino-openwebnet-new-openhab2-binding-ready-for-testing/43377/

This contribution is ready for review.

Closes #5195

Hilbrand commented on bundles/org.openhab.binding.openwebnet/README.md:
```diff
@@ -0,0 +1,248 @@
+- Contributors of this binding have no liability for any direct, indirect, incidental, special, exemplary, or consequential damage to things or people caused by using the binding connected to a real BTicino/Legrand (OpenWebNet) plant/system and its physical devices. The final user is the only responsible for using this binding in a real environment. See Articles 5. and 6. of [Eclipse Public Licen ...
+- The OpenWebNet protocol is maintained and Copyright by BTicino/Legrand. The documentation of the protocol if freely accessible for developers on the [MyOpen Community website - https://www.myopen-legrandgroup.com/developers](https://www.myopen-legrandgroup.com/developers/)
+- OpenWebNet, MyHOME, MyHOME_Play and Living Now are registered trademarks by BTicino/Legrand
+- This binding uses `openwebnet-lib 0.9.x`, an OpenWebNet Java lib partly based on [openwebnet/rx-openwebnet](https://github.com/openwebnet/rx-openwebnet) client library by @niqdev, modified to support:
```
If I read this correctly the library here is a modified version, but I don't see any links to the source of that modified library only to the original library?

mvalla replied:
Hi @Hilbrand , thanks for starting the review. The library used started as a fork of [openwebnet/rx-openwebnet](https://github.com/openwebnet/rx-openwebnet), but then there were too many missing parts (mainly the USB serial part and events handling) so I started deviating quite a bit from that lib.
Current status is that I have a quite messy lib that I did not publish yet on GitHub.
It would take 4/5 weeks at least for me to clean up the library, refactor some parts to be decent, update the binding to use the new library and do some tests.
This would mean to miss the OH 2.5.0 release planned in December, and that would be not good at all since many users are waiting for this binding to be part of the official release.
I propose the following: the current binding is merged with the current lib, then after 2.5.0 release I will add more devices (WHOs) and the new restructured openwebnet lib to communicate with openwebnet gateways. What do you think?
In any case I want to keep the lib separated as it is quite useful to have a OpenWebNet Java library with all the characteristics I added: nothing similar can be found on opensource and other projects may benefit from such a lib, not just OH. Current dependencies of my openwebnet lib are: slf4j-log4j12, nrjavaserial, com.google.guava, org.eclipse.jdt.annotation

mvalla replied:
Hi @Hilbrand, I think I addressed all your recommendations, correct?
If yes, would it be possible to continue your review of the binding?

mvalla replied:
The binding is now based on openwebnet4j lib bundle, available here: https://github.com/mvalla/openwebnet4j

Hilbrand commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java:
```diff
@@ -0,0 +1,217 @@
+        private boolean isBTicino = false;
+
+        private DeviceInfo(RemoteDevice device) {
+            logger.debug("+=== UPnP =========================================");
```
Please remove log lines with only text. Logging runs in parallel. So in a user system with lots of things these lines just appear randomly between other log messages and thus make them less useful.

mvalla replied:
I grouped the device discovery print in one single line, and removed other multiple lines logs

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/OpenWebNetHandlerFactory.java:
```diff
@@ -0,0 +1,107 @@
+        if (OpenWebNetBridgeHandler.SUPPORTED_THING_TYPES.contains(thing.getThingTypeUID())) {
+            logger.debug("==OWN:HandlerFactory== creating NEW BRIDGE Handler");
+            OpenWebNetBridgeHandler handler = new OpenWebNetBridgeHandler((Bridge) thing);
+            registerDiscoveryService(handler);
```
ok, modified in last commit

mvalla replied:
ok modified to use `ThingHandlerService`, removed `registerDiscoveryService `and `removeHandler `

wborn commented on bundles/org.openhab.binding.openwebnet/NOTICE:
```diff
@@ -0,0 +1,20 @@
+
+== Third-party Content
+
+Modified version of rx-openwebnet
```
Can you provide the sources of this modified version?
Seems the only way people can now modify/maintain this binding is by [decompiling the library](https://community.openhab.org/t/bticino-openwebnet-new-openhab2-binding-ready-for-testing/43377/1049?u=wborn).

mvalla replied:
The binding is now based on openwebnet4j lib bundle, available here: https://github.com/mvalla/openwebnet4j

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetAutomationHandler.java:
```diff
@@ -0,0 +1,401 @@
+
+    public OpenWebNetAutomationHandler(@NonNull Thing thing) {
+        super(thing);
+        logger.debug("==OWN:AutomationHandler== constructor");
```
Can these messages be replaced by using the debugger? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java:
```diff
@@ -0,0 +1,577 @@
+        gateway.subscribe(this);
+        if (gateway.isConnected()) { // gateway is already connected, device can go ONLINE
+            isGatewayConnected = true;
+            logger.info("==OWN== ------------------- ALREADY CONNECTED -> setting status to ONLINE");
```
Is logging the prefix `===OWN===` necessary? Logging includes already the class' name.

mvalla commented:
> There are many TODOs and questions in the code. Is this ready to be reviewed? I took a peek at 11 of 34 files.

Many thanks @fwolter for this partial review. I will take care of your change requests. 
In fact I am in the process of migrating the binding to use a new lib to communicate with the OpenWebNet protocol. This lib has been re-written from scratch by me and the code will be published.
To simplify code review I further limited the binding functionality to Lights (ON,OFF, Dimmer): other functionalities will be added after the first version of the binding is merged.
I will ping you when this new version is ready, so that you can continue the review.

mvalla commented on bundles/org.openhab.binding.openwebnet/README.md:
```diff
@@ -0,0 +1,248 @@
+
+| Category             | WHO          | Thing Type IDs                      | Description                                                 | Status           |
+| -------------------- | :----------: | :---------------------------------: | ----------------------------------------------------------- | ---------------- |
+| Gateway Management   | `13`         | `bus_gateway`                       | Any IP gateway supporting OpenWebNet protocol should work (e.g. F454 / MyHOMEServer1 / MH202 / F455 / MH200N, ...) | Successfully tested: F454, MyHOMEServer1, MyHOME_Screen10, F455, F452, F453AV, MH201, MH202, MH200N. Some connection stability issues/gateway resets reported with MH202  |
```
Ok, I can do that, but that would mean all current users (~1500) should change their configuration. If this is ok for maintainers, I will proceed with this change.

fwolter replied:
Ok, I think that's not worth it. I would leave it, then. Where did you get this number from?

mvalla replied:
It's an estimate from the number of clicks on the link in the main thread (Popular links) to the binding [README_beta](https://community.openhab.org/t/bticino-openwebnet-new-openhab2-binding-ready-for-testing/43377) file which contains install instructions

fwolter commented:
Ok, I like this decision. I recommend to close this PR and open a new one, when your new code, written from scratch is ready for review.

mvalla commented:
> Ok, I like this decision. I recommend to close this PR and open a new one, when your new code, written from scratch is ready for review.

The OpenWebNet lib is written from scratch, not the binding code which remains 95% the same.
Is opening a new PR really needed? I would not like to be pushed at the end of the long binding review queue... ;-)

mvalla commented on bundles/org.openhab.binding.openwebnet/README.md:
```diff
@@ -0,0 +1,248 @@
+
+| Channel Type ID (channel ID)        | Item Type     | Description                                                             | Read/Write |
+|--------------------------|---------------|-------------------------------------------------------------------------|:----------:|
+| `switch`                 | Switch        | To switch the device `ON` and `OFF`                                     |    R/W     |
+| `brightness`             | Dimmer        | To adjust the brightness value (Percent, `ON`, `OFF`)                   |    R/W     |
```
Some things have just On/Off, other have On/Off and Dimmer. So, as done in other bindings for example Hue binding, I had to define two different channels to handle all cases. Is this wrong?

fwolter replied:
Ok, I thought you defined an additional Switch Channel to be able to use a Switch in the sitemap for a Dimmer thing. I understand that there is a hardware, which cannot be dimmed, but switched. And there is another hardware, which supports both. In that case it's ok to stick with two different Channels.

fwolter commented:
Then, you should stay with this PR.

mvalla commented:
Hallo @fwolter. I addressed all your comments and more importantly, the binding is now based on the new lib bundle openwebnet4j, source code available here: https://github.com/mvalla/openwebnet4j.
(I am going to publish the bundle on bintray soon).

So this openwebnet binding first PR is now complete from my point of view and its review can now continue.

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetThingHandler.java:
```diff
@@ -0,0 +1,205 @@
+
+    @Override
+    public void dispose() {
+        logger.debug("dispose() for {}", getThing().getUID());
```
This is logged by the framework and can be removed.

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetThingHandler.java:
```diff
@@ -0,0 +1,205 @@
+    protected abstract String ownIdPrefix();
+
+    @SuppressWarnings("unchecked")
+    protected <U extends Quantity<U>> QuantityType<U> commandToQuantityType(Command command, Unit<U> defaultUnit) {
```
Is this method used anywhere or can it be removed?

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/OpenWebNetHandlerFactory.java:
```diff
@@ -0,0 +1,97 @@
+    private void registerDiscoveryService(OpenWebNetBridgeHandler bridgeHandler) {
+        OpenWebNetDeviceDiscoveryService deviceDiscoveryService = new OpenWebNetDeviceDiscoveryService(bridgeHandler);
+        bridgeHandler.deviceDiscoveryService = deviceDiscoveryService;
+        // deviceDiscoveryService.activate();
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java:
```diff
@@ -0,0 +1,229 @@
+
+        public static @Nullable BusGatewayId fromValue(String s) {
+            Optional<BusGatewayId> m = Arrays.stream(values()).filter(val -> s.equals(val.value)).findFirst();
+            if (m == null) {
```
`m` cannot be null at this point. The null check could be removed.

mvalla replied:
Changed to m.isPresent(), as m is an Optional

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java:
```diff
@@ -0,0 +1,229 @@
+    /**
+     * DeviceInfo bean to store device useful info (and log them)
+     */
+    // @NonNullByDefault is not necessary here
```
You could remove this comment.

mvalla replied:
I put it there to remind this is not necessary: one source code checker complains about subclass not being annotated, while the other gives compile error if you put @NonNullByDefault. Anyway I removed it.

fwolter replied:
There has been a bug in checkstyle that caused this false positive. It's fixed in OH3.

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java:
```diff
@@ -0,0 +1,229 @@
+                this.udn = identity.getUdn();
+                deviceLog += "\n| ID.UDN       : " + udn;
+                if (identity.getDescriptorURL() != null) {
+                    deviceLog += "\n| ID.DESC URL  : " + identity.getDescriptorURL();
+                    this.host = identity.getDescriptorURL().getHost();
+                }
+                deviceLog += "\n| ID.MAX AGE : " + identity.getMaxAgeSeconds();
+            }
+            deviceLog += "\n| --------------";
+            DeviceDetails details = device.getDetails();
+            if (details != null) {
+                ManufacturerDetails manufacturerDetails = details.getManufacturerDetails();
+                if (manufacturerDetails != null) {
+                    this.manufacturer = manufacturerDetails.getManufacturer();
+                    deviceLog += "\n| MANUFACTURER : " + manufacturer + " (" + manufacturerDetails.getManufacturerURI()
+                            + ")";
+                    if (manufacturer.toUpperCase().contains("BTICINO")) {
+                        this.isBTicino = true;
+                    }
+                }
+                ModelDetails modelDetails = details.getModelDetails();
+                if (modelDetails != null) {
+                    // Model Name | Desc | Number (Uri)
+                    this.modelName = modelDetails.getModelName();
+                    this.modelDescription = modelDetails.getModelDescription();
+                    this.modelNumber = modelDetails.getModelNumber();
+                    deviceLog += "\n| MODEL        : " + modelName + " | " + modelDescription + " | " + modelNumber
+                            + " (" + modelDetails.getModelURI() + ")";
+                }
+                if (isBTicino) {
+                    this.friendlyName = details.getFriendlyName();
+                    deviceLog += "\n| FRIENDLY NAME: " + friendlyName;
+                    this.serialNumber = details.getSerialNumber();
+                    deviceLog += "\n| SERIAL #     : " + serialNumber;
+                    deviceLog += "\n| BASE URL     : " + details.getBaseURL();
+                    deviceLog += "\n| UPC          : " + details.getUpc();
+                }
+            }
+            deviceLog += "\n+==================================================";
+            logger.debug(deviceLog);
```
This is quite a lot of code for logging. Can use the debugger instead?

mvalla replied:
It's both logging and recognising the fields of the found UPnP device.  After many tests with users I have found so many variants of info published via UPnP -> I have found very useful to log at debug level which info is delivered via UPnP and show even more info if the device is a BTicino one. There is no other easy way to show such info to the user, who can always snooze it setting log level to INFO

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java:
```diff
@@ -0,0 +1,229 @@
+                Map<String, Object> properties = new HashMap<>(4);
+                String label = "BUS Gateway (" + thingId.getId().split("-")[0] + ")";
+                try {
+                    label = ((!("".equals(devInfo.friendlyName))) ? devInfo.friendlyName : "NO_NAME");
```
```suggestion
                    label = devInfo.friendlyName.isEmpty() ? "NO_NAME" : devInfo.friendlyName;
```

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java:
```diff
@@ -0,0 +1,229 @@
+                try {
+                    label = ((!("".equals(devInfo.friendlyName))) ? devInfo.friendlyName : "NO_NAME");
+                    label = label + " (" + devInfo.modelName + ", " + devInfo.modelNumber + ", " + devInfo.host + ")";
+                } catch (Exception e) {
```
Which type of exception do you expect here?

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java:
```diff
@@ -0,0 +1,229 @@
+            if (udn != null) {
+                idString = udn.getIdentifierString();
+                if (idString != null) {
+                    BusGatewayId gwId = BusGatewayId.fromValue(idString.split("-")[1]);
```
Are you sure there is always a dash in the ID? Better add a length check to the array. Same for below.

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/OpenWebNetDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,164 @@
+        super(SEARCH_TIME);
+        bridgeHandler = handler;
+        bridgeUID = handler.getThing().getUID();
+        logger.debug("Constructor for bridge: '{}'", bridgeUID);
```
Can this message be replaced by using the debugger?

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/OpenWebNetDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,164 @@
+        String thingLabel = OpenWebNetBindingConstants.THING_LABEL_GENERIC_DEVICE;
+        Who deviceWho = Who.UNKNOWN;
+        switch (deviceType) {
+            case ZIGBEE_ON_OFF_SWITCH: {
```
The brackets are not necessary.
```suggestion
            case ZIGBEE_ON_OFF_SWITCH:
```

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/resources/ESH-INF/thing/BusGateway.xml:
```diff
@@ -0,0 +1,54 @@
+				<required>true</required>
+			</parameter>
+
+			<parameter name="port" type="integer">
```
You could add min/max values here.

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/resources/ESH-INF/thing/BusGateway.xml:
```diff
@@ -0,0 +1,54 @@
+			</parameter>
+
+			<parameter name="passwd" type="text">
+				<context>password-create</context>
```
Since you are entering an existing password and not creating a new one:
```suggestion
				<context>password</context>
```

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/resources/ESH-INF/thing/BusGateway.xml:
```diff
@@ -0,0 +1,54 @@
+
+			<parameter name="port" type="integer">
+				<label>Port</label>
+				<description>OpenWebNet gateway port (default: 20000)</description>
```
Can you add a `<default>` tag?

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/OpenWebNetDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,164 @@
+                break;
+            }
+            default: {
+                logger.warn("Device type {} is not supported, default to GENERIC device (WHERE={})", deviceType, where);
```
If I understand correctly, your're adding generic devices you can't control or receive information for. What's the purpose of these?

mvalla replied:
Sometime devices are there but not recognised correctly, other times since they are not powered they are found by the gateway but their type is not recognised. By creating such GENERIC Things helps the user and the programmer to understand better the situation. I think something similar happens for the ZigBee binding, where unrecognised devices are still found but not controllable.

fwolter commented on bundles/org.openhab.binding.openwebnet/src/main/resources/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,40 @@
+	</channel-type>
+
+
+	<channel-type id="dimmerLevel">
```
Is this Channel used anywhere?

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetLightingHandler.java:
```diff
@@ -0,0 +1,376 @@
+            }
+        } else {
+            logger.warn("Cannot handle command {} for thing {}", command, getThing().getUID());
+            return;
```
I think this Warning is useful to the user to understand the command has just sent to an item (for example via REST API) is not handled by this thing. The effect is to warn the user about the unsupported command.

fwolter replied:
I fully aggree. I meant the return statement.

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetLightingHandler.java:
```diff
@@ -0,0 +1,376 @@
+        if (brH != null && brH.isBusGateway()) {
+            channelID = CHANNEL_SWITCH;
+        } else {
+            if (WhereZigBee.UNIT_02.equals(((WhereZigBee) (msg.getWhere())).getUnit())) {
```
There is a check if gateway is BUS or not; in case gw is ZigBee, msgs received from openwebnet protocol must be addressed using WhereZigBee format otherwise the parser in the openwebnet4j lib would fail parsing the msg and drop it. Anyway I made the code more readable

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetThingHandler.java:
```diff
@@ -0,0 +1,205 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("initialize() thing={}", thing.getUID());
```
I tried to use WARN only for unexpected cases, remove almost all INFO logging, just left some DEBUG level login where I found particularly useful to understand better what is going on...

fwolter replied:
Ok, but initialization of a Thing is already logged by the framework. So, this would be logged twice.

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetThingHandler.java:
```diff
@@ -0,0 +1,205 @@
+            OpenGateway gw = handler.gateway;
+            if (gw != null && !gw.isConnected()) {
+                logger.warn("Gateway is NOT connected, setting thing={} to OFFLINE", thing.getUID());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
```
Yes I put bridge thing to offline when I receive onConnectionError/onDisconnected callbacks, so also things beloging to bridge should go offline. I can remove this updateStatus.

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetThingHandler.java:
```diff
@@ -0,0 +1,205 @@
+     */
+    protected void handleMessage(BaseOpenMessage msg) {
+        ThingStatus ts = getThing().getStatus();
+        if (ThingStatus.ONLINE != ts && ThingStatus.REMOVING != ts && ThingStatus.REMOVED != ts) {
+            updateStatus(ThingStatus.ONLINE);
```
I could have a message coming from the openwebnet network, received as callback by Bridge handler with `OpenWebNetBridgeHandler.onEventMessage`, and delivered to this ThingHandler *exactly* while the Thing is in REMOVING status.
With:
`ThingStatus.REMOVING != ts`
I avoid the risk to make the thing become wrongly ONLINE again (it's also a not allowed transition from `ThingStatus.REMOVING`).
Then the removing process will complete in `OpenWebNetThing.dispose()`, where the device id is unregistered from bridge handler so that new openwebnet messages will not be delivered to this disposed thing anymore.
It probably added this check because during testing this happened at least in few cases.

fwolter replied:
You're right. This can happen. I didn't expect this behavior.

mvalla commented:
Hi @fwolter I think addressed all your change requests (or explained motivation for the few ones I did not change). 
What is the next step?
I have seen compilation with TravisCI fails because the `openwebnet4j ` lib is searched on Maven Central:
`Could not resolve dependencies for project org.openhab.addons.bundles:org.openhab.binding.openwebnet:jar:2.5.7-SNAPSHOT: Could not find artifact com.github.openwebnet4j:openwebnet4j:jar:0.1.0 in google-maven-central
`
Should all libs linked by OH2 be published on Central?
I remember examples of libs used by other bindings published just on JCenter...
Thanks for any hint!

fwolter commented:
> Hi @fwolter I think addressed all your change requests (or explained motivation for the few ones I did not change).
> What is the next step?
> I have seen compilation with TravisCI fails because the `openwebnet4j ` lib is searched on Maven Central:
> `Could not resolve dependencies for project org.openhab.addons.bundles:org.openhab.binding.openwebnet:jar:2.5.7-SNAPSHOT: Could not find artifact com.github.openwebnet4j:openwebnet4j:jar:0.1.0 in google-maven-central `
> Should all libs linked by OH2 be published on Central?
> I remember examples of libs used by other bindings published just on JCenter...
> Thanks for any hint!

Before answering your question in detail, please let me ask you some general questions. Could you share your reasons why you extracted your code into a library? Is this library used elsewhere? Are there any other contributors beside you, maintaining the library?

mvalla commented:
> Could you share your reasons why you extracted your code into a library? Is this library used elsewhere? Are there any other contributors beside you, maintaining the library?

The openwebnet part was in a separate lib jar since the beginning, I made it now as a separate module (openwebnet4j, OSGi) since jars in the `/lib` folder are not accepted anymore starting OH2.5 and in OH3.
As stated already in a [previous comment to @Hilbrand](https://github.com/openhab/openhab-addons/pull/6213#discussion_r341844274), I think it's very useful to keep the lib separated from binding as there is no Java openwebnet lib available with all the characteristics I added: support for USB ZigBee gateway, handling of notifications from gw, support for HMAC authentication, and others.
Other Java projects could benefit from the openwebnet4j lib and, in general, is good sw engineering practice to separate protocol/message handling (in this case OpenWebNet) from application (in this case this OH binding).
Hope this clarifies the reason I spent *quite an effort* to make the openwebnet part as a separate component (java OSGi lib).

Since the lib is new, for the moment it's not used elsewhere and there are no other contributors. However I received interest from some developers that offered to contribute actively to the project adding new functionalities, o support new devices they have connected to BTicino system.

mvalla commented:
> Could you share your reasons why you extracted your code into a library? Is this library used elsewhere? Are there any other contributors beside you, maintaining the library?

The openwebnet part was in a separate lib jar since the beginning, I made it now as a separate module (openwebnet4j, OSGi) since jars in the `/lib` folder are not accepted anymore starting OH2.5 and in OH3.
As stated already in a [previous comment to @Hilbrand](https://github.com/openhab/openhab-addons/pull/6213#discussion_r341844274), I think it's very useful to keep the lib separated from binding as there is no Java openwebnet lib available with all the characteristics I added: support for USB ZigBee gateway, handling of notifications from gw, support for HMAC authentication, and others.
Other Java projects could benefit from the openwebnet4j lib and, in general, is good sw engineering practice to separate protocol/message handling (in this case OpenWebNet) from application (in this case this OH binding).
Hope this clarifies the reason I spent *quite an effort* to make the openwebnet part as a separate component (java OSGi lib).

Since the lib is new, for the moment it's not used elsewhere and there are no other contributors. However I received interest from some developers that offered to contribute actively to the project adding new functionalities, o support new devices they have connected to BTicino system.

fwolter commented:
I agree that it's good practice to separate code into modules to have well-defined interfaces and I understood that you put much effort into doing this. The reason why I'm asking is, if you're not available anymore (we won't hope so), it'll be hard for the OH community to maintain this binding. Because when something needs to be changed in your lib, it needs to be forked and the fork will be used only in OH, then. I wouldn't have ask you if we know there is a community that maintains the lib and that requests are getting processed reliably upstream. So, this is hard for us maintainers. Here is my suggestion: Integrate the code into the OH binding, release the lib in parallel and look for some contributors. If there's been a community built around the lib, the OH binding makes use of it.

mvalla commented:
I surely understand the issue of making sure the binding can be maintained.
However there are surely examples of few other bindings with similar approach: where the protocol part has been correctly separated in another module. There are also a few examples where there is no real community around the lib used in a binding.
The reasons I want to keep the lib separate are: visibility/independence of the lib, copyright ownership, and reuse of code. There is also practical reason: if someone improves the protocol code in the binding, then someone else should port the improvement back to the lib, and this should happen for each improvement. This really does not make sense and would not simply happen, marking they lib evolve only within OH. Dependencies have been introduced to modularize software and OH itself is highly modular thanks to OSGi.
What are the other options available or used in three past?

J-N-K commented:
@mvalla Since I already know (and worked on) at least two bindings where external libraries are not maintained anymore and we were forced to fork to fix issues (moquette for the embedded MQTT broker and hap-java for HomeKit) let me add my two cents.

In one case (moquette) the code is nearly unmaintainable. We are in a situation where we either need to find a replacement, rewrite the whole broker based on another library or drop the bundle. Since no one steps up to do that work, we‘ll probably drop it.

In the other case (hap-java) luckily someone added everything we needed and fixed bugs. We now maintain our own fork, which is the only reason why the addon still works. The latest official release can‘t be used with recent iOS versions. We are currently discussing if we can add the code with our additions as third-party-content in our codebase. This is very difficult due to license issues.

There are some more examples where not everything is freely available and some native code is binary-only. This leads to the situation where the bundle works on some platforms but not all. And there is nothing we can do about it. This is very unfortunate.

Of course, we can‘t and will not try to force you in one direction or the other. But please re-think what is the best we can achieve. I would very much appreciate seeing your extensive work in our repository.

fwolter commented:
Here is another suggestion: You keep the lib and the openHAB code separate; If there's no activity on the library repo and no interest by anyone, we could move it to openHAB. Likewise, if you aren't maintaining the binding anymore, we will have the right to move the code to the openHAB repo (and use our namespace and license headers with it).

WDYT?

mvalla commented:
Thank you @fwolter and @J-N-K for the comments and examples.

The moquette case shows that even a populated  dev community does not guarantee to keep a lib error free or to become abandonwere.
The hap-java example shows that the last suggestion from @fwolter is always a valid option : fork in case the project is not maintained, eventually integrate code if license allows it (this is surely the case of MIT to be included into EPL-2.0, keeping license+namespace+copyright+attribution, I do not understand the mentioned licensing problems there).

In conclusion I would go for the last option suggested by @fwolter : keep the lib separate for now and consider the several options available in case the lib by its own does not evolve (which of course I would not hope).

So if this is agreed , what would be the next step for this binding PR ?

kaikreuzer commented:
> eventually integrate code if license allows it (this is surely the case of MIT to be included into EPL-2.0, keeping license+namespace+copyright+attribution, I do not understand the mentioned licensing problems there).

@mvalla: The general issue here is that we cannot integrate external code easily into openHAB, since we would have to keep license, copyright, etc. exactly in place. That's why we'd like to ask for your upfront approval to move the code into openHAB just like classes that are part of the binding (i.e. as EPLv2 with the standard openHAB license header) - clearly only for the situation that your lib is no longer maintained, while the openHAB binding further evolves. Hope this makes clear what @fwolter suggested!

mvalla commented:
Thanks @kaikreuzer for the further clarification. I think, in case the external openwebnet4j lib should become abandoned, a copyright transfer can be done. In practice moving the lib code to the OH repo using the OH copyright. But let’s hope this is not going to happen: if there is interest in developing further the binding, there should be interest in contributing to the lib as first thing.

kaikreuzer commented:
> But let’s hope this is not going to happen

I'm fully with you here! 😎 Arrangement above is merely meant as a safety net, which we hopefully won't need.

kaikreuzer commented:
JCenter is normally enough as we consider it as a source for dependencies in openHAB - hence the Jenkins PR build also succeeded. It should be fine to ignore the failing Travis build, which might be due to the incremental build nature.

kaikreuzer commented:
@Hilbrand and @wborn Could you please check whether your comments are addressed?

Hilbrand commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java:
```diff
@@ -0,0 +1,566 @@
+     *
+     * @param listener to receive device found notifications
+     */
+    public synchronized void searchDevices() {
```
I would have expected most of this discovery code to be in the discovery service. To keep code better separated. Or is there a reason the code is here in the handler?

mvalla replied:
`searchDevices `and `onNewDevice`/`onDiscoveryCompleted `are here because it's `OpenWebNetBridgeHandler ` that owns the communication to the BTicino gateway (by using openwebnet protocol), through which device discovery is performed.

mvalla commented on bundles/org.openhab.binding.openwebnet/README.md:
````diff
@@ -0,0 +1,198 @@
+}
+```
+
+## Disclaimer
````
changed to ## Notes and left only link to OpenWebNet protocol docs

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java:
```diff
@@ -0,0 +1,566 @@
+     * Init a BUS/SCS gateway based on config properties
+     */
+    private @Nullable OpenGateway initBusGateway() {
+        if (getConfig().get(CONFIG_PROPERTY_HOST) != null) {
```
ok done. I had to add 2 new classes, so not sure it makes code simpler in this case. Anyway... I did copy from KNX/HUE bindings.

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetLightingHandler.java:
```diff
@@ -0,0 +1,376 @@
+                    }
+                }
+                if (!(command instanceof PercentType)) {
+                    updateState(channel, new PercentType(Lighting.levelToPercent(newWhatInt)));
```
ok I improved the logic around dimmer control to avoid duplicate state updates

mvalla commented:
I completed all changes requested by @Hilbrand in his 2nd review and updated the openwebnet4j lib to 0.2.0 to include OPEN/HMAC password support. This PR is again ready from my side. All checks have passed.

martinvw commented:
@Hilbrand if you did a complete review, I would say this has two reviewers. If you feel it needs another round let us know otherwise feel free to merge.

mvalla commented:
Thanks for merging this pull request. I will advertise it in the relevant thread.
